### ISSUE TITLE
Add instance of any dynamic type to database as part of binding

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#2477]
Basically, implement the functionality described in this https://github.com/Starcounter/Starcounter/issues/2473#issuecomment-69332832 (minus that the new ```Car``` is assigned the auto-created model; that will be in a seperate issue).

This will be done in branch **next**.

**Comment by per-samuelsson** [comment]
The status of this right now is that we are struggling with a problem regarding how auto-created types are dealt with during unload/reload. The rest is basically in place.

The design is that metadata has been extended, so that it keeps track of what user classes has been auto-instantiated to create dynamic types. The metadata keeps a reference from a given class (e.g. ```Car```) to it's auto-created type (e.g. ```CarModel```).

The binding consume this metadata. If it finds such a reference, the target object is validated and no new type is created.

The problem is that when an unload occur, metadata is not part of the data. Hence, the auto created type will be part of the unload data, but **the tracking reference wont be**. Or, in the case of how the SQL regression test work, the metadata remain in the database (i.e. everything is deleted except metadata tables) and then the binding fail, because the binding runs before the reload, and hence the tracking reference will point to an object that don't yet exist.

I'm currently trying different approaches to solve this, but I still havent found any satisfying.

**Comment by k-rus** [comment]
One approach is to forbid deleting the auto-created instance by user. The only way to delete the instance will be to delete table, which instances derive from it.
In my opinion users should be able to modify only instances, which was created by them, i.e., by application code.

**Comment by per-samuelsson** [comment]
> One approach is to forbid deleting the auto-created instance by user.

Yes, this is already planned as part of the implementation. But it doesn't solve this problem.

> In my opinion users should be able to modify only instances, which was created by them, i.e., by application code.

That might make some things easier, but then again, I think it's sub-optimal considering the usage of types. Also, I'm not sure either it would solve the problem.

After testing out different approaches, and considered how the old rebuild handled this case (it was a somewhat similar thing there), I still have no real solution unfortunatelly. The only things I have come up with that is 100% sure is

a) if we want to preserve the metadata that indicates what types have been auto-created by Starcounter, we can't handle it as the "other" metadata; instead, we must handle it as I first modelled it, part of "user code space", to assure it ends up in the dump.

b) the reload is a bit tricky since it's designed to run at any time; this is not how it was done in the previous implementation. In versions < 2.x, reload was part of the binding/initialization of the database. Hence, it was possible to delay some things to after a reload, and do up-front initialization logic there. This would have solved the current case. (But maybe thats a bad idea for other reasons, considering how the reload now actually have been implemented).

**Comment by per-samuelsson** [comment]
For the record, one idea both @k-rus and I have had is to consider if it's possible not to do any auto-creating of types as part of binding, but instead delay it so that it is assured only when an instance of a concept is created.

Hence, the ```CarModel``` instance would not be auto-created until the first ```Car``` is created, illustrated by this modification of [this original code](https://github.com/Starcounter/Starcounter/issues/2473#issuecomment-69332832):

```C#
[Database] public class Car {
  [Type]
  public CarModel Model;
}

[Database] public class CarModel {
  public string ModelName;
}

class Program {
  static void Main() {
    Console.WriteLine(Db.SQL<Car>("SELECT * FROM Car").First);
    Console.WriteLine(Db.SQL<CarModel>("SELECT * FROM CarModel").First);
    Console.WriteLine(new Car().Model);
    Console.WriteLine(Db.SQL<CarModel>("SELECT * FROM CarModel").First);
  }
}

/* Output: 
      null
      null
      [model]
      [model]
*/
```

Notice here that the first query to retreive any ```CarModel``` result in **null**; not until there has actually been a ```Car``` **instantiated**, a ```CarModel``` is assured.

The obvious drawback here is that it will affect runtime performance (extra check during creation of instances), but this might be able to isolate only to cases where dynamic types are actually in use.

Another, possible more serious, drawback is that types are created at unknown times, and also that it will cause strange effects when there are multiple transactions auto-creating types (and all but one will fail because of conflict).

**Comment by per-samuelsson** [comment]
We have decided on a meeting today - @dan31, me, @k-rus, @Starcounter-Jack - that we'll keep the implementation as is, and instead let it be a known limitation in **next** that the rebuild does not work. The reasoning is that we will still need to adapt it for other things.

Hence, I will disable the tests in **next** that includes rebuilding of databases and try to go on from there.

**Comment by per-samuelsson** [comment]
This is now implemented. I'm closing this, continuing with #2493.
